feat: merge user supplied and default plugin configs

naseemkullah opened the pull request:
fixes #616

## Which problem is this PR solving?

By merging user supplied config two levels deep, a user can now only configure the diff
from the default auto enablement of installed plugins, rather than
having to explicitly re enable all plugins if at least one of their
configurations require are edited.

Furthermore, user supplied plugins not listed in default plugins are
implicitly enabled.


## Short description of the changes

1. User supplied config is checked for any default plugin configs, those are merged with the default config of default plugins.

2. User supplied plugins which are not part of the default plugins list are implicitly enabled, but can be disabled with `enabled: false` property.

3. The merged user supplied plugins are then merged with the default plugins to include any default plugins that may not have been configured by the user.

dyladan commented:
Can you please add tests

naseemkullah commented:
> Can you please add tests

Trying to test brought to light that simple use of object spread only merged the user supplied and default configs at the top level. Seeing that these configs go 2 levels deep **I installed `deepmerge` for this purpose.**

However I am tearing my hair out trying to write a test to test the deep merge.

![image](https://user-images.githubusercontent.com/24660299/80149325-7860ab80-8584-11ea-80f8-e9f710fe37b0.png)


As seen above, I'd like to run an `assert` type test for the HttpPlugin's
   
```json
_config: {
    enabled: true,
    path: '@opentelemetry/plugin-http',
    ignoreUrls: [ 'foo' ]
  }
```

to demonstrate that `enabled` and `path` properties remain intact despite the user supplied config being only for `ignoreUrls`.... but it seems to be inaccessible:

![image](https://user-images.githubusercontent.com/24660299/80149488-c1186480-8584-11ea-97ca-999f00a05430.png)


Please advise @dyladan @mayurkale22 

Also worth noting: plugin-http was added as a dev dependency.

naseemkullah commented:
We could reduce scope and revert to object spread, any user configured plugins would need to be fully configured (enabled, path), non-configured default plugins would implicitly load as intended, if this is preferred.

dyladan commented on packages/opentelemetry-node/package.json:
```diff
@@ -42,6 +42,7 @@
   },
   "devDependencies": {
     "@opentelemetry/context-base": "^0.6.1",
+    "@opentelemetry/plugin-http": "^0.6.1",
```
Why was this added?

naseemkullah replied:
to load the http plugin in test case

naseemkullah replied:
(but test case is not complete yet as per https://github.com/open-telemetry/opentelemetry-js/pull/980#issuecomment-618671197)

markwolff replied:
Would it make sense to add a mock plugin instead?
https://github.com/open-telemetry/opentelemetry-js/blob/23677a16f60f6dea89f449607b5549f711d6ff99/packages/opentelemetry-core/test/trace/BasePlugin.test.ts#L60

naseemkullah replied:
I seem to already be using the provided mock plugin, and this is not required. Removed!
Thanks.

dyladan commented on packages/opentelemetry-node/package.json:
```diff
@@ -64,6 +65,7 @@
     "@opentelemetry/context-async-hooks": "^0.6.1",
     "@opentelemetry/core": "^0.6.1",
     "@opentelemetry/tracing": "^0.6.1",
+    "deepmerge": "^4.2.2",
```
Definitely do not want to depend on some new random dependency for such simple functionality

naseemkullah replied:
do we want deep merge or shallow merge?

naseemkullah replied:
seems healthy fwiw https://www.npmjs.com/package/deepmerge

markwolff replied:
+1 to not adding a dependency. I think a shallow merge is fine here, and then we can rely on each plugin handling its own config merge if needed.

dyladan replied:
I also think shallow merge is fine. The "default" configuration is just an empty configuration for all plugins anyways so there is no need to merge with it. Default settings are handled within the plugin itself.

naseemkullah replied:
sgtm, but what I still do not understand, if we just shallow merge, with regards to default settings being handled within the pluging itself if:

```
        plugins: {
          'http': {
            ignoreUrls: ['foo'],
          },
        }
```

would result in http being enabled and it's path being `'@opentelemetry/plugin-http' ` or not?

dyladan replied:
No you will need to make the `path` property required. Or you will need to implement the config merge, but no need to make it a true deep merge. One level should be enough.

naseemkullah replied:
gotcha!

naseemkullah replied:
Alright 2 level merge working, I've added a todo to automatically enable unless explicitly disabled, user configured non-default plugins, but that could be a subsequent PR.

naseemkullah replied:
pardon the forced pushes, should be done now.

naseemkullah commented:
This is ready for review @dyladan but seems to fail for what appears to be unrelated tests, anything to do with https://github.com/open-telemetry/opentelemetry-js/commit/e4e71b8b49d516bb6902ee6562f7ca1eba689182#diff-d94ef5d9aa7453023c0b440fbb324e91 ?

markwolff commented on packages/opentelemetry-node/src/NodeTracerProvider.ts:
```diff
@@ -39,7 +39,34 @@ export class NodeTracerProvider extends BasicTracerProvider {
+      } else {
+        mergedUserSuppliedPlugins[pluginName] = config.plugins[pluginName];
+        // enable user-supplied plugins unless explicitly disabled
+        if (mergedUserSuppliedPlugins[pluginName].enabled === undefined) {
```
Is `enabled` a required plugin config? If not, I think it would be a good field to add.

naseemkullah replied:
By not requiring it, we would be able to enable by default. Is it preferred to require `enabled` be explicitly set, or default to `enabled: true` when not set?

naseemkullah replied:
Just to clarify my intent: that `enabled` is a plugin config parameter, but if left unset, it would default to true. So if someone brings their plugin `horse` and specifies the path, it will automatically enable without them having to add `enabled: true` explicitly.

naseemkullah replied:
Seems like it is not required:

https://github.com/open-telemetry/opentelemetry-js/blob/b1fe2c81b9842dca18dba7db7d87c15fc95820b6/packages/opentelemetry-api/src/trace/instrumentation/Plugin.ts#L56-L60

But it also seems to default to `true` but I'm not sure why/how that gets lost, and why I am defaulting it to true for any user supplied plugin.

naseemkullah commented:
bump

naseemkullah commented:
bump

dyladan commented:
>I think we should add a specific test that assert that the default options are not overriden by the user one to be sure we dont introduce a regression in the future

Default options should be overridden by user options.

vmarchaud commented:
@dyladan Sorry i meant the fact that the user pass its custom options override *all* default options

dyladan commented:
@naseemkullah sorry this has been such a long process. I agree with @vmarchaud we should probably have a test that if user supplies config for `http`, the `grpc` plugin for instance does not have it's config mangled or overridden.

naseemkullah commented:
np thanks for the feedback @dyladan @vmarchaud , I'll work on getting the test and let you know if any blockers come up (fetching the plugin config and being able to check it in a test for `grpc` for example).

naseemkullah commented:
Alright I added a test by checking if grpc plugin would successfully load without any grpc related config in a situation with user supplied plugin config.

Furthermore, not sure how to incorporate this in test, but i console logged `{plugins}`:

```
{
  plugins: [
    SimpleModulePlugin {
      _tracerName: undefined,
      _tracerVersion: undefined,
      moduleName: 'simple-module',
      _moduleExports: [Object],
      _tracer: [Tracer],
      _logger: NoopLogger {},
      _internalFilesExports: {},
      _config: [Object]
    },
    SimpleModulePlugin {
      _tracerName: undefined,
      _tracerVersion: undefined,
      moduleName: 'supported-module',
      supportedVersions: [Array],
      _moduleExports: [Object],
      _tracer: [Tracer],
      _logger: NoopLogger {},
      _internalFilesExports: {},
      _config: [Object]
    },
    GrpcPlugin {
      _tracerName: '@opentelemetry/plugin-grpc',
      _tracerVersion: '0.8.0',
      moduleName: 'grpc',
      version: '1.24.2',
      supportedVersions: [Array],
      _internalFilesList: [Object],
      _basedir: '/open-telemetry/opentelemetry-js/packages/opentelemetry-node/node_modules/grpc',
      _config: [Object],
      _moduleExports: [Object],
      _tracer: [Tracer],
      _logger: NoopLogger {},
      _internalFilesExports: [Object]
    }
  ]
}
```

dyladan commented on packages/opentelemetry-node/package.json:
```diff
@@ -42,12 +42,14 @@
   },
   "devDependencies": {
     "@opentelemetry/context-base": "^0.8.0",
+    "@opentelemetry/plugin-grpc": "^0.8.0",
```
You don't need to load the actual grpc plugin for the config. You just need to test that your merge config function doesn't overwrite options for other plugins.

```typescript
const defaultConfig = {
  module1: {
    enabled: true,
    path: 'testpath'
  },
  module2: {
    enabled: true,
    path: 'testpath2'
  }
};

const userConfig = {
  module2: {
    path: 'userpath'
  }
}

const merged = mergeConfig(defaultConfig, userConfig);

// assert merged.module1.path is testpath
// assert merged.module2.path is userpath
```

naseemkullah replied:
Thanks @dyladan, I did not think about factoring out `mergePlugins` (which I call it as such since the types it merges and returns are `Plugins` rather than plugin configs.)

I've exported it for the sole purpose of importing it in the test suite.

dyladan commented on packages/opentelemetry-node/test/NodeTracerProvider.test.ts:
```diff
@@ -280,3 +278,33 @@ describe('NodeTracerProvider', () => {
+    },
+  };
+
+  const mergedPlugins = mergePlugins(defaultPlugins, userPlugins);
```
This line should be inside the test

dyladan commented on packages/opentelemetry-node/test/NodeTracerProvider.test.ts:
```diff
@@ -269,3 +278,32 @@ describe('NodeTracerProvider', () => {
 });
+
+describe('mergePlugins', () => {
+  const defaultPlugins = {
```
Please add a test that if a user sets some config, the default config is not mangled.

```typescript
  const defaultPlugins = {
    module1: {
      enabled: true,
      path: 'testpath',
    },
    module2: {
      enabled: true,
      path: 'testpath2',
    },
  };

  const userPlugins = {
    module2: {
      path: 'userpath',
    },
  };

  const mergedPlugins = mergePlugins(defaultPlugins, userPlugins);

  // assert defaultPlugins.module2.path is testpath2
```

dyladan commented:
Please stop force pushing changes until the review process is complete

naseemkullah commented:
I've addressed your comments and added a `nonDefaultModule` as well, while I am trying to cover more ground some of these tests may be gratuitous/redundant, WDYT @dyladan

naseemkullah commented on packages/opentelemetry-node/src/NodeTracerProvider.ts:
```diff
@@ -55,3 +60,39 @@ export class NodeTracerProvider extends BasicTracerProvider {
+        ...userSuppliedPlugins[pluginName],
+      };
+    } else {
+      mergedUserSuppliedPlugins[pluginName] = userSuppliedPlugins[pluginName];
```
Thanks, I always felt a little off about that line, not I know why!

naseemkullah commented on packages/opentelemetry-node/src/NodeTracerProvider.ts:
```diff
@@ -55,3 +60,39 @@ export class NodeTracerProvider extends BasicTracerProvider {
+  let mergedUserSuppliedPlugins: Plugins = {};
+
+  for (const pluginName in userSuppliedPlugins) {
+    if (DEFAULT_INSTRUMENTATION_PLUGINS.hasOwnProperty(pluginName)) {
```
That is much cleaner, thanks @obecny !

naseemkullah replied:
Refactored 👍

naseemkullah commented:
Sorry I did not want to force push right now 🙈

mayurkale22 commented on packages/opentelemetry-node/src/NodeTracerProvider.ts:
```diff
@@ -55,3 +60,32 @@ export class NodeTracerProvider extends BasicTracerProvider {
+  defaultPlugins: Plugins,
+  userSuppliedPlugins: Plugins
+): Plugins {
+  let mergedUserSuppliedPlugins: Plugins = {};
```
`s/let/const`?

naseemkullah replied:
Thanks, isn't `gts` supposed to catch this sort of thing? 🤔

mayurkale22 commented on packages/opentelemetry-node/src/NodeTracerProvider.ts:
```diff
@@ -55,3 +60,32 @@ export class NodeTracerProvider extends BasicTracerProvider {
+ * Then merge the results with the default plugins.
+ * @returns 2-layer deep merge of default and user supplied plugins.
+ */
+export function mergePlugins(
```
Usually I don't prefer to export function only for tests. Perhaps, we can do something like this:

```
export const TEST_ONLY = {
  mergePlugins
};
```
WDYT? Feel free to ignore if you don't agree.

naseemkullah replied:
I'm not sure what best practice is, but it did feel a little off to export something strictly to test it, will try your above mentioned approach, thanks!

dyladan replied:
Think I would prefer to make it a private member of the class which can be called in tests using the `instance["_privateFunction"]()` syntax

naseemkullah replied:
til, sgtm

naseemkullah replied:
done via https://github.com/open-telemetry/opentelemetry-js/pull/980/commits/03d6a482fb5312fad86f13e071891b56b493e61b

dyladan commented on packages/opentelemetry-node/src/NodeTracerProvider.ts:
```diff
@@ -59,33 +59,30 @@ export class NodeTracerProvider extends BasicTracerProvider {
+   * Then merge the results with the default plugins.
+   * @returns 2-layer deep merge of default and user supplied plugins.
+   */
+  mergePlugins(defaultPlugins: Plugins, userSuppliedPlugins: Plugins): Plugins {
```
make this a private method.

naseemkullah replied:
Thanks @dyladan !

mayurkale22 commented:
@naseemkullah Would be nice if we can document this behavior somewhere in the [README section](https://github.com/open-telemetry/opentelemetry-js/tree/master/packages/opentelemetry-node) (in the separate PR). WDYT?

naseemkullah commented:
> @naseemkullah Would be nice if we can document this behavior somewhere in the [README section](https://github.com/open-telemetry/opentelemetry-js/tree/master/packages/opentelemetry-node) (in the separate PR). WDYT?

💯 I can include it in this PR, np. Where would be a suitable place to talk about plugin config?

naseemkullah commented:
> > @naseemkullah Would be nice if we can document this behavior somewhere in the [README section](https://github.com/open-telemetry/opentelemetry-js/tree/master/packages/opentelemetry-node) (in the separate PR). WDYT?
> 
> 💯 I can include it in this PR, np. Where would be a suitable place to talk about plugin config?

I guess opentelemetry-node package README itself

mayurkale22 commented:
> > @naseemkullah Would be nice if we can document this behavior somewhere in the [README section](https://github.com/open-telemetry/opentelemetry-js/tree/master/packages/opentelemetry-node) [](/open-telemetry/opentelemetry-js/tree/HEAD@{2020-05-21T00:30:12Z}/packages/opentelemetry-node) (in the separate PR). WDYT?
> 
> 💯 I can include it in this PR, np.

Awesome, thank you so much. This is applicable for both Node and Web, maybe we can write in both the places.

naseemkullah commented:
> > > @naseemkullah Would be nice if we can document this behavior somewhere in the [README section](https://github.com/open-telemetry/opentelemetry-js/tree/master/packages/opentelemetry-node)  (in the separate PR). WDYT?
> > 
> > 
> > 💯 I can include it in this PR, np.
> 
> Awesome, thank you so much. This is applicable for both Node and Web, maybe we can write in both the places.

My pleasure! I've added a first pass of the doc in the node README, please let me know what you think.

I am puzzled as to how this affects the WebTracer though?

mayurkale22 commented:
@dyladan would be nice if we can include this in `0.8.2`

mayurkale22 commented:
> > > > @naseemkullah Would be nice if we can document this behavior somewhere in the [README section](https://github.com/open-telemetry/opentelemetry-js/tree/master/packages/opentelemetry-node)  (in the separate PR). WDYT?
> > > 
> > > 
> > > 100 I can include it in this PR, np.
> > 
> > 
> > Awesome, thank you so much. This is applicable for both Node and Web, maybe we can write in both the places.
> 
> My pleasure! I've added a first pass of the doc in the node README, please let me know what you think.
> 
> I am puzzled as to how this affects the WebTracer though?

@obecny do you think we need to update Web Readme?

obecny commented:
> @obecny do you think we need to update Web Readme?
For web you always have to provide the plugins, so if I understand tis correctly there aren't any changes to the web here

naseemkullah commented:
I am having issues running lint and lint:fix locally, might have something to do with recent migration to eslint, any ideas @dyladan @mayurkale22 ? 

```
Oops! Something went wrong! :(

ESLint: 6.8.0.

ESLint couldn't find the plugin "eslint-plugin-header".

(The package "eslint-plugin-header" was not found when loaded as a Node module from the directory "/Users/naseemullah/repos/github.com/open-telemetry/opentelemetry-js/packages/opentelemetry-node".)

It's likely that the plugin isn't installed correctly. Try reinstalling by running the following:

    npm install eslint-plugin-header@latest --save-dev
```

dyladan commented:
I would first try to do a full clean and rebuild as the dependencies are all very different

```sh
lerna clean -y && rm -rf node_modules && npm install
```

naseemkullah commented:
> ```shell
> ```shell
> lerna clean -y && rm -rf node_modules && npm install
> ```
> 
> 
>       
> ```

tried from the `opentelemetry-node` folder with no luck trying from root folder...

dyladan commented:
> tried from the `opentelemetry-node` folder with no luck trying from root folder...

This should be run from the root folder. If you're on windows you may need to split it into separate commands:

```
lerna clean -y
# somehow delete node_modules idk windows :P
npm install
```

naseemkullah commented:
> > tried from the `opentelemetry-node` folder with no luck trying from root folder...
> 
> This should be run from the root folder. If you're on windows you may need to split it into separate commands:
> 
> ```
> lerna clean -y
> # somehow delete node_modules idk windows :P
> npm install
> ```

Ah ok, so from root folder should work.

I've been living Windows free since my last enterprise gig a few years back!
Apart from when I help my dad with his computer :D

naseemkullah commented:
Seem to be getting errors on non related code:

```

/root/project/packages/opentelemetry-context-zone-peer-dep/src/types.ts
  26:25  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  28:6   warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  31:25  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
  33:6   warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any

✖ 14 problems (5 errors, 9 warnings)

```

dyladan commented:
Those are warnings not errors that you posted. The warnings don't fail the build

mayurkale22 commented:
> ~Seem to be getting errors on non related code:~
> 
> ```
> 
> /root/project/packages/opentelemetry-context-zone-peer-dep/src/types.ts
>   26:25  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
>   28:6   warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
>   31:25  warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
>   33:6   warning  Unexpected any. Specify a different type  @typescript-eslint/no-explicit-any
> 
> ✖ 14 problems (5 errors, 9 warnings)
> ```

https://github.com/open-telemetry/opentelemetry-js/issues/1093 is created to address these warnings. Please ignore it for now.

naseemkullah commented:
> Those are warnings not errors that you posted. The warnings don't fail the build

Ah right but still seems to be the case,  message updated to include more context

naseemkullah commented:
There are errors as well, `npm run lint` finally works again locally, changes in question are linted and I think this is good to go if we ignore the errors from zoneContextMgr

dyladan commented:
hmm. the question is why does it not break the build for others...

mayurkale22 commented:
```sh
$ git clean -dfx
$ npm install
$ npm run lint
```
I tried these commands locally and everything is working well for me.

naseemkullah commented:
now

```
lerna ERR! npm run lint exited 1 in '@opentelemetry/context-async-hooks'
lerna ERR! npm run lint stdout:

> @opentelemetry/context-async-hooks@0.8.2 lint /root/project/packages/opentelemetry-context-async-hooks
> eslint . --ext .ts


/root/project/packages/opentelemetry-context-async-hooks/src/AsyncHooksContextManager.ts
   46:20  error    Expected a `const` instead of a literal type assertion                                                                                                                                                                                                                                                                                                                                                              @typescript-eslint/prefer-as-const
   47:11  error    Expected a `const` instead of a literal type assertion                                                                                                                                                                                                                                                                                                                                                              @typescript-eslint/prefer-as-const
   48:13  error    Expected a `const` instead of a literal type assertion                                                                                                                                                                                                                                                                                                                                                              @typescript-eslint/prefer-as-const
   49:24  error    Expected a `const` instead of a literal type assertion                                                                                                                                                                                                                                                                                                                                                              @typescript-eslint/prefer-as-const
   50:28  error    Expected a `const` instead of a literal type assertion                                                                                                                                                                                                                                                                                                                                                              @typescript-eslint/prefer-as-const
   95:37  warning  Unexpected any. Specify a different type                                                                                                                                                                                                                                                                                                                                                                            @typescript-eslint/no-explicit-any
  144:35  error    Don't use `Function` as a type. The `Function` type accepts any function-like value.
It provides no type safety when calling the function, which can be a common source of bugs.
It also accepts things like class declarations, which will throw at runtime as they will not be called with `new`.
If you are expecting the function to accept certain arguments, you should explicitly define the function shape  @typescript-eslint/ban-types
  146:44  error    Don't use `{}` as a type. `{}` actually means "any non-nullish value".
- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.
- If you want a type meaning "any value", you probably want `unknown` instead                                                                                                                                                                @typescript-eslint/ban-types
  159:30  warning  Unexpected any. Specify a different type                                                                                                                                                                                                                                                                                                                                                                            @typescript-eslint/no-explicit-any
  205:67  error    Don't use `Function` as a type. The `Function` type accepts any function-like value.
It provides no type safety when calling the function, which can be a common source of bugs.
It also accepts things like class declarations, which will throw at runtime as they will not be called with `new`.
If you are expecting the function to accept certain arguments, you should explicitly define the function shape  @typescript-eslint/ban-types
  206:12  warning  Missing return type on function                                                                                                                                                                                                                                                                                                                                                                                     @typescript-eslint/explicit-module-boundary-types
  206:28  error    Don't use `{}` as a type. `{}` actually means "any non-nullish value".
- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.
- If you want a type meaning "any value", you probably want `unknown` instead                                                                                                                                                                @typescript-eslint/ban-types
  227:15  error    Don't use `Function` as a type. The `Function` type accepts any function-like value.
It provides no type safety when calling the function, which can be a common source of bugs.
It also accepts things like class declarations, which will throw at runtime as they will not be called with `new`.
If you are expecting the function to accept certain arguments, you should explicitly define the function shape  @typescript-eslint/ban-types
  229:12  warning  Missing return type on function                                                                                                                                                                                                                                                                                                                                                                                     @typescript-eslint/explicit-module-boundary-types
  229:28  error    Don't use `{}` as a type. `{}` actually means "any non-nullish value".
- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.
- If you want a type meaning "any value", you probably want `unknown` instead                                                                                                                                                                @typescript-eslint/ban-types
  250:15  error    Don't use `Function` as a type. The `Function` type accepts any function-like value.
It provides no type safety when calling the function, which can be a common source of bugs.
It also accepts things like class declarations, which will throw at runtime as they will not be called with `new`.
If you are expecting the function to accept certain arguments, you should explicitly define the function shape  @typescript-eslint/ban-types
  254:12  warning  Missing return type on function                                                                                                                                                                                                                                                                                                                                                                                     @typescript-eslint/explicit-module-boundary-types
  254:28  error    Don't use `{}` as a type. `{}` actually means "any non-nullish value".
- If you want a type meaning "any object", you probably want `Record<string, unknown>` instead.
- If you want a type meaning "any value", you probably want `unknown` instead                                                                                                                                                                @typescript-eslint/ban-types

✖ 18 problems (13 errors, 5 warnings)
  5 errors and 0 warnings potentially fixable with the `--fix` option.
```

dyladan commented:
Should be fixed with #1094

dyladan commented:
I think it is most likely a versioning discrepancy as the lint versions are not pinned in CI. #1081 also fixes this shortcoming.

rlibm commented:
Tryign to confirm that the regression in the plugin load behaviour is intentional or not. My intention is to have a few plugins loaded for instrumentation. I have been loading an array of `plugins` to the `NodeTraceProvider` with code like (yes, I have my own set of default plugins which covers `http`, `https`, `express`, and some proprietary modules):

```js
    const plugins = composePlugins(options.plugins || defaultPlugins, options.pluginOptions || {});
    const config = {
      plugins,
      sampler: ALWAYS_SAMPLER,
      traceParams: {
        numberOfEventsPerSpan: 64,
        numberOfAttributesPerSpan: 64
      },
      logger
    };

    this.#provider = new NodeTracerProvider(config);
```

We are intentionally not instrumenting some other libraries, for example, `ioredis`.

With this change, we are getting errors on startup:

```
ERROR PluginLoader#load: could not load plugin @opentelemetry/plugin-ioredis of module ioredis. Error: Cannot find module '@opentelemetry/plugin-ioredis'
Require stack:
- /home/rolf/bmix/platapi-core/node_modules/@opentelemetry/node/build/src/instrumentation/PluginLoader.js
- /home/rolf/bmix/platapi-core/node_modules/@opentelemetry/node/build/src/NodeTracerProvider.js
- /home/rolf/bmix/platapi-core/node_modules/@opentelemetry/node/build/src/index.js
......
```

I see now that all the `DEFAULT_INSTRUMENTATION_PLUGINS` plugins are enabled.

I would like to remove all of these default plugins, or disable them all, but I cannot see an exposed way to do that.

3 questions:

1. was it intentional to change the behaviour this way, and will it stay this way?
2. how do I get the list of default plugins so I can run through them and disable them all?
3. is there a better way?

naseemkullah commented:
> Trying to confirm that the regression in the plugin load behaviour is intentional or not. My intention is to have a few plugins loaded for instrumentation. I have been loading an array of `plugins` to the `NodeTraceProvider` with code like (yes, I have my own set of default plugins which covers `http`, `https`, `express`, and some proprietary modules):
> 
> ```js
>     const plugins = composePlugins(options.plugins || defaultPlugins, options.pluginOptions || {});
>     const config = {
>       plugins,
>       sampler: ALWAYS_SAMPLER,
>       traceParams: {
>         numberOfEventsPerSpan: 64,
>         numberOfAttributesPerSpan: 64
>       },
>       logger
>     };
> 
>     this.#provider = new NodeTracerProvider(config);
> ```
> 
> We are intentionally not instrumenting some other libraries, for example, `ioredis`.
> 
> With this change, we are getting errors on startup:
> 
> ```
> ERROR PluginLoader#load: could not load plugin @opentelemetry/plugin-ioredis of module ioredis. Error: Cannot find module '@opentelemetry/plugin-ioredis'
> Require stack:
> - /home/rolf/bmix/platapi-core/node_modules/@opentelemetry/node/build/src/instrumentation/PluginLoader.js
> - /home/rolf/bmix/platapi-core/node_modules/@opentelemetry/node/build/src/NodeTracerProvider.js
> - /home/rolf/bmix/platapi-core/node_modules/@opentelemetry/node/build/src/index.js
> ......
> ```
> 
> I see now that all the `DEFAULT_INSTRUMENTATION_PLUGINS` plugins are enabled.
> 
> I would like to remove all of these default plugins, or disable them all, but I cannot see an exposed way to do that.
> 
> 3 questions:
> 
> 1. was it intentional to change the behaviour this way, and will it stay this way?
> 2. how do I get the list of default plugins so I can run through them and disable them all?
> 3. is there a better way?

Hi @rlibm 

1. It was, and as such I think it will stay that way, provided your ux can be improved.
2. I know of 3 ways:
  - [README](https://github.com/open-telemetry/opentelemetry-js/blob/master/README.md#node-plugins) 
  - [source code](https://github.com/open-telemetry/opentelemetry-js/blob/deb35707b768537619ecf79225476f177c157df2/packages/opentelemetry-node/src/config.ts#L29-L40)
3. Logs when starting your process to see what plugins that were not explicitly enabled are trying to load.

With that said we could surely reduce the severity of

> ERROR PluginLoader#load: could not load plugin @opentelemetry/plugin-ioredis of module ioredis. Error: Cannot find module '@opentelemetry/plugin-ioredis'

to something like

> WARN PluginLoader#load: could not load plugin @opentelemetry/plugin-ioredis of module ioredis. Error: Cannot find module '@opentelemetry/plugin-ioredis', if instrumentation of this module is undesired, please ignore.

Or add yet another config option to disable default plugins. @dyladan @mayurkale22 thoughts comments or corrections to the above?

rlibm commented:
The specific issue I have is this code is being put in a library that wraps opentelemetry, and fortunately for me, I noticed the error before it went further. I am fortunate that the first system I tested it on had ioredis installed, and threw that error, and I am also fortunate that the `plugin-ioredis` was not included in the dependencies.

Otherwise the ioredis and other plugins may have been silently, and unexpectedly loaded, and caused significant increases in tracing output, and possibly issues in down-stream automated anomoly detection, etc.

In effect, the risk I have is 2-fold:

1. I have to continuously monitor whether the default plugin list has changed in each code-release of opentelemetry, disabling plugins as they arrive....
2. I need positive assurance that only the plugins I want to load, are loaded.

I would recommend some option on the NodeTracingProvider that bypasses the merge, and uses the supplied plugins as-configured.

Automatic merging with unknown content is a bad solution in enterprise environments.

naseemkullah commented:
> The specific issue I have is this code is being put in a library that wraps opentelemetry, and fortunately for me, I noticed the error before it went further. I am fortunate that the first system I tested it on had ioredis installed, and threw that error, and I am also fortunate that the `plugin-ioredis` was not included in the dependencies.
> 
> Otherwise the ioredis and other plugins may have been silently, and unexpectedly loaded, and caused significant increases in tracing output, and possibly issues in down-stream automated anomoly detection, etc.
> 
> In effect, the risk I have is 2-fold:
> 
> 1. I have to continuously monitor whether the default plugin list has changed in each code-release of opentelemetry, disabling plugins as they arrive....
> 2. I need positive assurance that only the plugins I want to load, are loaded.
> 
> I would recommend some option on the NodeTracingProvider that bypasses the merge, and uses the supplied plugins as-configured.
> 
> Automatic merging with unknown content is a bad solution in enterprise environments.

Is it fortune to not have `plugin-ioredis` in the dependencies or a desire to not instrument `ioredis`? 😛 

In any case, I think a `mergeDefault` option (enabled by default plz) would suit you/enterprise environments. How's about a PR?

rlibm commented:
We have the desire to not instrument ioredis. for cloud.ibm.com the redis volumes are huge. Inadvertently enabling it would be .... problematic.

Side-Note, we're migrating from OpenCensus to OpenTelemetry. The OpenTelemetry is really close to production (yes, we know this library is not yet out of Beta).

I will work on a PR, but we natively use JavaScript, I don't have tooling set up for .ts as such. May take a while.

naseemkullah commented:
@rlibm Could you elaborate on the downsides of instrumenting a huge redis dataset? Do you think it would add latency and if so would it be possible to do some benchmarking? 

That is exciting that IBM Cloud is migrating to OTEL! 🚀 

FWIW We are starting to use this quite a bit in production at Transit, but we are a startup 😄 

Great to hear you are up for a PR!!! 👍  TS is the best!!!

naseemkullah commented:
ooh and one more thing, don't be a stranger if you encounter any time wasters/blockers when working on the PR https://gitter.im/open-telemetry/opentelemetry-node
